src/Avalonia.Controls/Generators/TabItemContainerGenerator.cs

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) The Avalonia Project. All rights reserved.
+// Licensed under the MIT license. See licence.md file in the project root for full license information.
+
+namespace Avalonia.Controls.Generators
+{
+    using Avalonia.Controls.Primitives;
+
+    public class TabItemContainerGenerator : ItemContainerGenerator<TabItem>
+    {
+        public TabItemContainerGenerator(TabControl owner)
+            : base(owner, ContentControl.ContentProperty, ContentControl.ContentTemplateProperty)
+        {
+            Owner = owner;
+        }
+
+        public new TabControl Owner { get; }
+
+        protected override IControl CreateContainer(object item)
+        {
+            var tabItem = (TabItem)base.CreateContainer(item);
+
+            tabItem.ParentTabControl = Owner;
+
+            if (tabItem.HeaderTemplate == null)
+            {
+                tabItem[~HeaderedContentControl.HeaderTemplateProperty] = Owner[~ItemsControl.ItemTemplateProperty];
+            }
+
+            if (tabItem.Header == null)
+            {
+                if (item is IHeadered headered)
+                {
+                    if (tabItem.Header != headered.Header)
```

Comment:
Is this check needed? We've established that `tabItem.Header` is null on line 29.

Reply:
Should be safe to remove that check 👍